source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -900,6 +901,293 @@ void MSETRequest::onChildResponse(Common::Redis::RespValuePtr&& value, uint32_t
+  return handler;
+}
+
+bool requiresValue(const std::string& arg) {
+    return (arg == "count" || arg == "match" || arg == "type");
+}
+
+// Generic function to build an array with bulkstring
+void addBulkString(Common::Redis::RespValue& requestArray, const std::string& value) {
+    Common::Redis::RespValue element;
+    element.type(Common::Redis::RespType::BulkString);
+    element.asString() = value;
+    requestArray.asArray().emplace_back(std::move(element));
+}
+
+SplitRequestPtr ScanRequest::create(Router& router, Common::Redis::RespValuePtr&& incoming_request,
+                                    SplitCallbacks& callbacks, CommandStats& command_stats,
+                                    TimeSource& time_source, bool delay_command_latency,
+                                    const StreamInfo::StreamInfo& stream_info) {
+
+  // SCAN looks like: SCAN cursor [MATCH pattern] [COUNT count] [TYPE type]
+  // Ensure there are at least two args to the command or it cannot be scanned.
+  // Also the number of arguments should be in even number, otherwise command is invalid
+  if (incoming_request->asArray().size() < 2 || incoming_request->asArray().size() % 2 != 0) {
+    onWrongNumberOfArguments(callbacks, *incoming_request);
+    command_stats.error_.inc();
+    return nullptr;
+  }
+
+  std::string key = std::string();
+  int32_t shard_idx = 0; 
+  int32_t numofRequests= 3;
+  std::string count = "2";
+
+  Common::Redis::RespValue requestArray;
+  requestArray.type(Common::Redis::RespType::Array);
+
+  // Getting the right cursor before sending request
+  // We are appending 4 digit custom value as shard index to the cursor returned from the previous scan request
+  // If it is a first request, the received cursor will be sent directly without any preprocessing
+  // Ex: If cursor length is more than 4 -> last four digits will be considered as index and the remaining will be taken as cursor
+  // If cursor length is less than or equal to 4 -> all digits will be considered as cursor
+  std::string cursor = incoming_request->asArray()[1].asString();
+  if (cursor.length() > 4) {
+    std::string index = cursor.substr(cursor.length() - 4);
+    cursor = cursor.substr(0, cursor.length() - 4);
+    shard_idx = std::stoi(index);
+  }
+
+  // Completely reconstructing the request to add/modify count since we can't override the incoming array directly
+  // Add the command and cursor to the request array
+  addBulkString(requestArray, "SCAN");
+  addBulkString(requestArray, cursor);
+
+  // Iterate over the arguments modify count if necessary
+  // We are setting 1000 as the default count value
+  for (size_t i = 2; i < incoming_request->asArray().size(); ++i) {
+      std::string arg = incoming_request->asArray()[i].asString();
+      addBulkString(requestArray, arg);
+
+      // Check if the argument requires a value
+      if (requiresValue(arg)) {
+          if (arg == "count") {
+            addBulkString(requestArray, count);
```

Comment:
least of the user provided count and 1000 has to be given
fix this

Reply:
Fixed this